redis/src/entra_id.rs

```diff
@@ -0,0 +1,951 @@
+//! It supports multiple credential types including DeveloperToolsCredential, service principals,
+//! managed identities, as well as custom `TokenCredential` implementations.
+//!
+//! # Features
+//!
+//! - **Multiple Authentication Flows**: Service principals, managed identities, and custom `TokenCredential` implementations
+//! - **Automatic Token Refresh**: Background token refresh with configurable policies
+//! - **Retry Logic**: Robust error handling with exponential backoff
+//! - **Async Support**: Full async/await support for non-blocking operations
+//!
+//! # Example
+//!
+//! ```rust,no_run
+//! use redis::{Client, EntraIdCredentialsProvider, RetryConfig};
+//!
+//! # async fn example() -> redis::RedisResult<()> {
+//! // Create the credentials provider using the DeveloperToolsCredential
+//! let mut provider = EntraIdCredentialsProvider::new_developer_tools()?;
+//! provider.start(RetryConfig::default());
+//!
+//! // Create Redis client with credentials provider
+//! let client = Client::open_with_credentials_provider(
+//!     "redis://your-redis-instance.com:6380",
+//!     provider
+//! )?;
+//!
+//! // Use the client to get a multiplexed connection
+//! let mut con = client.get_multiplexed_async_connection().await?;
+//! redis::cmd("SET")
+//!     .arg("my_key")
+//!     .arg(42i32)
+//!     .exec_async(&mut con)
+//!     .await?;
+//! let result: Option<String> = redis::cmd("GET")
+//!     .arg("my_key")
+//!     .query_async(&mut con)
+//!     .await?;
+//! # Ok(())
+//! # }
+//! ```
+
+use crate::auth::BasicAuth;
+use crate::auth::StreamingCredentialsProvider;
+use crate::auth_management::credentials_management_utils;
+use crate::errors::{ErrorKind, RedisError};
+use crate::types::RedisResult;
+use crate::RetryConfig;
+use azure_core::credentials::{AccessToken, Secret, TokenCredential};
+use azure_identity::{
+    ClientCertificateCredential, ClientCertificateCredentialOptions, ClientSecretCredential,
+    ClientSecretCredentialOptions, DeveloperToolsCredential, DeveloperToolsCredentialOptions,
+    ManagedIdentityCredential, ManagedIdentityCredentialOptions,
+};
+use futures_util::{Stream, StreamExt};
+use std::pin::Pin;
+use std::sync::{Arc, Mutex, RwLock};
+use time::OffsetDateTime;
+use tokio::sync::mpsc::Sender;
+
+/// The default Redis scope for Azure Managed Redis
+pub const REDIS_SCOPE_DEFAULT: &str = "https://redis.azure.com/.default";
+
+/// A client certificate in PKCS12 (PFX) that can be used for client certificate authentication.
+///
+/// The certificate data should be base64-encoded PKCS12 content.
+/// If the PKCS12 archive is password-protected, provide the password via `password`.
+#[derive(Debug, Clone)]
+pub struct ClientCertificate {
+    /// Base64-encoded PKCS12 certificate data
+    pub base64_pkcs12: String,
+    /// The certificate's password if any
+    pub password: Option<String>,
+}
+
+type Subscriptions = Vec<Arc<Sender<RedisResult<BasicAuth>>>>;
```

Comment:
I don't think that the `Arc` here is necessary, the sender is already cheaply clonable.